src/engine/surface_collision.c

```diff
@@ -704,7 +704,7 @@ s32 find_water_level(s32 x, s32 z) { // TODO: Allow y pos
 
             // If the location is within a water box and it is a water box.
             // Water is less than 50 val only, while above is gas and such.
-            if (loX < x && x < hiX && loZ < z && z < hiZ && val < 50) {
+            if (loX <= x && x <= hiX && loZ <= z && z <= hiZ && val < 50) {
```

Comment:
What does this do exactly? Vanilla doesn't use `<=`.

Reply:
in vanilla one of these is a float, for some reason in hacker it was changed to both ints. this results in a unit-sized gap between adjacent waterboxes (most notably the waterboxes in JRB) which causes mario to upwarp occasionally when travelling between them. in vanilla this was the size of a floating point imprecision so it basically never happened but now it's much easier to trigger.

Reply:
In that case I'd assume it technically makes more sense to leave `hiX` and `hiZ` as <, but it probably doesn't really matter anyway if we leave it like this.